## 1.Tìm hiểu về đề tài:

### 1.1. Giới thiệu về đề tài:

- **Tên cuộc thi:** Recruit Restaurant Visitor Forecasting
- **Mô tả:** 
    - Việc dự đoán sẽ dựa vào các thông tin về vị trí (kinh độ, vĩ độ) của nhà hàng, các ngày lễ trong năm, thông tin dự báo thời tiết ngày hôm đó và dữ liệu đặt chỗ thông qua trang web đặt chỗ của nhà hàng và dữ liệu khách đã đến trong cơ sở dữ liệu của nhà hàng.
    - **Output:** số lượng khách đến nhà hàng
- **Link kaggle:** https://www.kaggle.com/competitions/recruit-restaurant-visitor-forecasting/overview

### 1.2. Lí do chọn đề tài

Trong bối cảnh chuyển đổi số diễn ra mạnh mẽ, việc ứng dụng khoa học dữ liệu và học máy (Machine Learning) vào các bài toán thực tế ngày càng trở nên quan trọng, đặc biệt trong lĩnh vực kinh doanh và dịch vụ. Một trong những bài toán có giá trị ứng dụng cao là dự báo nhu cầu khách hàng, giúp doanh nghiệp tối ưu vận hành và nâng cao hiệu quả kinh doanh.

Cuộc thi “Recruit Restaurant Visitor Forecasting” trên nền tảng Kaggle đặt ra bài toán dự đoán số lượng khách ghé thăm nhà hàng trong tương lai dựa trên dữ liệu lịch sử về lượt khách, đặt chỗ, thông tin nhà hàng và lịch ngày lễ. Đây là một bài toán dự báo chuỗi thời gian (Time Series Forecasting) mang tính thực tiễn cao, phản ánh đúng các thách thức mà doanh nghiệp trong ngành nhà hàng – dịch vụ đang phải đối mặt.

Đề tài được lựa chọn vì những lý do sau:

Thứ nhất, tính ứng dụng thực tế cao.
Việc dự đoán chính xác số lượng khách giúp nhà hàng chủ động trong việc chuẩn bị nguyên vật liệu, bố trí nhân sự, giảm thiểu lãng phí và nâng cao chất lượng phục vụ. Do đó, kết quả của bài toán không chỉ mang ý nghĩa học thuật mà còn có thể áp dụng trực tiếp vào hoạt động kinh doanh thực tế.

Thứ hai, phù hợp để nghiên cứu và thực hành các kỹ thuật học máy hiện đại.
Bài toán yêu cầu xử lý dữ liệu theo thời gian, khai thác các yếu tố như ngày trong tuần, ngày lễ, xu hướng theo mùa và hành vi đặt chỗ của khách hàng. Điều này tạo điều kiện để áp dụng các kỹ thuật quan trọng như:

Phân tích và xử lý chuỗi thời gian

Feature engineering từ dữ liệu thực tế

So sánh và đánh giá các mô hình dự báo

Qua đó giúp người học củng cố kiến thức lý thuyết và nâng cao kỹ năng triển khai mô hình.

Thứ ba, dữ liệu phong phú, đa nguồn và gần với dữ liệu thực tế.
Dataset của cuộc thi bao gồm nhiều bảng dữ liệu khác nhau như thông tin nhà hàng, lịch sử khách ghé thăm, dữ liệu đặt chỗ và lịch ngày lễ. Việc kết hợp các nguồn dữ liệu này phản ánh đúng đặc trưng của các bài toán dữ liệu trong thực tế, giúp rèn luyện khả năng tiền xử lý, làm sạch và tích hợp dữ liệu.

Thứ tư, phù hợp với định hướng học tập và nghiên cứu trong lĩnh vực Khoa học dữ liệu.
Đề tài cho phép mở rộng theo nhiều hướng như so sánh các mô hình dự báo, cải tiến đặc trưng đầu vào hoặc đánh giá tác động của các yếu tố bên ngoài đến lượng khách. Điều này rất phù hợp để phát triển thành đồ án môn học, đồ án tốt nghiệp hoặc dự án cá nhân nhằm xây dựng portfolio.

Từ những lý do trên, đề tài “Dự báo số lượng khách nhà hàng – Recruit Restaurant Visitor Forecasting” được lựa chọn nhằm kết hợp giữa lý thuyết và thực tiễn, đồng thời nâng cao năng lực phân tích dữ liệu và xây dựng mô hình học máy cho các bài toán dự báo trong đời sống thực tế.

## Ý tưởng để giải quyết bài toán:
- Giải quyết về vấn đề dữ liệu:
    - Dữ liệu cửa hàng
        - Số lượng khách hàng đến cửa hàng trong quá khứ: `air_visit_data.csv`
        - Thời gian trong dataset:  `date_info.csv`
        - Thông tin cửa hàng đính kèm với với trạm thời tiết gần đó: `air_store_info`
        - chuẩn bị tập test `submission.csv`
    - Dữ liệu thời tiết
        - Dữ liệu thời tiết của 1663 trạm
- Tiền xử lí dữ liệu
    - Xử lí giá trị ngoại lai
    - Tính trọng số theo cấp số nhân cho số lượng khác hàng, cùng với một số đặc trưng khác
- Huấn luyện mô hình
    - Mã hóa dữ liệu
    - Chia dữ liệu train, test
    - Huấn luyện mô hình
    - Hiển thị kết quả

## 2. Giải quyết vấn đề

In [35]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

In [1]:
import pandas as pd
import numpy as np
import lightgbm as lgb
from tqdm import tqdm
from sklearn.model_selection import KFold
import warnings

### 2.1. Dữ liệu nhà hàng

#### 2.1.1. File `air_visit_data.csv`

In [37]:
path1 = 'data/air_visit_data.csv'
air_visit = pd.read_csv(path1)
air_visit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252108 entries, 0 to 252107
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   air_store_id  252108 non-null  object
 1   visit_date    252108 non-null  object
 2   visitors      252108 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 5.8+ MB


- Thông qua việc đọc dữ liệu ta thấy rằng: file chứ các cột `air_store_id`, `visit_date`, `visitors`
- Ý nghĩa của các cột dữ liệu: 
    - **Với cột dữ liệu `air_store_id`:**
        - Đây là cột dữ liệu cung cấp id của cửa hàng, là một dãy kí tự có độ dài 20
        - Trạng thái: cột chứa 252108 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: `air_store_id` có kiểu dữ liệu là object
    - **Với cột dữ liệu `visit_date`:**
        - Đây là cột dữ liệu chứa thời gian có khách của cửa hàng, chứa ngày tháng năm
        - Trạng thái: cột chứa 252108 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: `visit_date` có kiểu dữ liệu là object
    - **Với cột dữ liệu `visitors`:**
        - Đây là cột dữ liệu mà nhóm phải dự đoán để đưa ra kết quả, là các số nguyên dương
        - Trang thái: cột chứa 252180 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: `visitors` có kiểu dữ liệu là int64

In [38]:
air_visit.head(10)

air_store_id  visit_date  visitors
0  air_ba937bf13d40fb24  2016-01-13        25
1  air_ba937bf13d40fb24  2016-01-14        32
2  air_ba937bf13d40fb24  2016-01-15        29
3  air_ba937bf13d40fb24  2016-01-16        22
4  air_ba937bf13d40fb24  2016-01-18         6
5  air_ba937bf13d40fb24  2016-01-19         9
6  air_ba937bf13d40fb24  2016-01-20        31
7  air_ba937bf13d40fb24  2016-01-21        21
8  air_ba937bf13d40fb24  2016-01-22        18
9  air_ba937bf13d40fb24  2016-01-23        26

- **Ý tưởng để xử lí file `visit_data.csv`:** 
    - Chuyển dữ liệu `visit_date` kiểu object sang kiểu thời gian chuẩn: nhóm sử dụng to_datetime() để làm việc đó
    - Làm liền mạch chuỗi dữ liệu thời gian: ở bảng dữ liệu phía trên, chúng ta có thể thấy rằng dữ liệu chỉ hiển thị những ngày có khách, vì vậy nhóm đã phải dùng lệnh `resample('1d')` để có thể tạo ra một chuỗi dữ liệu thời gian liên tục, các ngày thiểu - không có khách vẫn được thêm vào
    - Định dạng thời gian theo kiểu YYYY-mm-dd: bằng dt.strftime(%Y-%m-%d)
    - Thay dữ liệu trống: Vốn dĩ những ngày được thêm vào đều không có giá trị, nên nhóm quyết định thay thế bằng 0
    - Thêm nhãn nhận biết: nhóm đã thêm nhãn `was_nil`. Với giá trị `False` thì ta biết được đó là dữ liệu của hệ thống, còn với giá trị `True` thì là dữ liệu do chúng ta thêm vào
- **Output:** cung cấp cho ta một file với dữ liệu thời gian liền mạch

In [39]:
air_visit.index = pd.to_datetime(air_visit['visit_date'])
air_visit = air_visit.groupby('air_store_id').apply(lambda data: data['visitors'].resample('1d').sum(), include_groups=False).reset_index()
air_visit['visit_date'] = air_visit['visit_date'].dt.strftime('%Y-%m-%d') #chuyen dang YYYY-mm-dd
air_visit.replace(0, np.nan, inplace=True)
air_visit['was_nil'] = air_visit['visitors'].isnull()
air_visit.fillna({'visitors': 0}, inplace = True) #ignore

Dữ liệu file `air_visit_data.csv` sau khi xử lí là:

In [40]:

air_visit.head(10)

air_store_id  visit_date  visitors  was_nil
0  air_00a91d42b08b08d9  2016-07-01      35.0    False
1  air_00a91d42b08b08d9  2016-07-02       9.0    False
2  air_00a91d42b08b08d9  2016-07-03       0.0     True
3  air_00a91d42b08b08d9  2016-07-04      20.0    False
4  air_00a91d42b08b08d9  2016-07-05      25.0    False
5  air_00a91d42b08b08d9  2016-07-06      29.0    False
6  air_00a91d42b08b08d9  2016-07-07      34.0    False
7  air_00a91d42b08b08d9  2016-07-08      42.0    False
8  air_00a91d42b08b08d9  2016-07-09      11.0    False
9  air_00a91d42b08b08d9  2016-07-10       0.0     True

##

#### 2.1.2 `File date_info.csv`

In [41]:
date_info = pd.read_csv('data/date_info.csv')
date_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   calendar_date  517 non-null    object
 1   day_of_week    517 non-null    object
 2   holiday_flg    517 non-null    int64 
dtypes: int64(1), object(2)
memory usage: 12.2+ KB


- Thông qua việc xem dữ liệu file `date_info.csv` gồm 3 cột `calendar_date`, `day_of_week` và `holiday_flag`
- Ý nghĩa của các cột dữ liệu: 
    - **Với cột dữ liệu `calendar_date`:**
        - Đây là cột dữ liệu cung cấp thời gian 
        - Trạng thái: cột chứa 517 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: `calendar_date` có kiểu dữ liệu là object
    - **Với cột dữ liệu `day_of_week`:**
        - Đây là cột dữ liệu cung cấp cho chúng ta các thứ trong tuần
        - Trạng thái: cột chứa 517 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: `day_of_week` có kiểu dữ liệu là object
    - **Với cột dữ liệu `holiday_flg`:**
        - Đây là cột dữ liệu cung cấp nhãn về ngày đó có phải là ngày lễ hay không. Với giá trị là 1 thì đây là ngày lễ, ngược lại giá trị 0 thì không phải là ngày lễ
        - Trạng thái: cột chứa 517 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: `holiday_flg` có kiểu dữ liệu là int64

In [42]:
date_info.head()

calendar_date day_of_week  holiday_flg
0    2016-01-01      Friday            1
1    2016-01-02    Saturday            1
2    2016-01-03      Sunday            1
3    2016-01-04      Monday            0
4    2016-01-05     Tuesday            0

- **Ý tưởng để xử lí file date_info.csv:**
    - Đổi tên 2 cột `calendar_date` và `holiday_flg` thành `visit_date` và `is_holiday` để có thể tiện dùng
    - Tạo thêm 2 cột mới: nhóm tạo thêm `prev_holiday` và `after_holiday` để xét ngày trước và sau của ngày đang xét là ngày lễ, giá trị của 2 cột đều giống với cột `is_holiday`
- **Ouput:** ta sẽ được file với dữ liệu cho biết rằng ngày đó có trong kì nghỉ dài hạn hay không

In [43]:
date_info.rename(columns = {'calendar_date' : 'visit_date', 'holiday_flg' : 'is_holiday'}, inplace = True)
date_info['prev_holiday'] = date_info['is_holiday'].shift().fillna(0)
date_info['after_holiday'] = date_info['is_holiday'].shift(-1).fillna(1)

Đây là file `date_info.csv` sau khi được xử lí:

In [44]:
date_info.head()

visit_date day_of_week  is_holiday  prev_holiday  after_holiday
0  2016-01-01      Friday           1           0.0            1.0
1  2016-01-02    Saturday           1           1.0            1.0
2  2016-01-03      Sunday           1           1.0            0.0
3  2016-01-04      Monday           0           1.0            0.0
4  2016-01-05     Tuesday           0           0.0            0.0

#### 2.1.3 `File air_store_info.csv`

Với file `air_store_info.csv`, nhóm đã sử dụng bảng đã được file đã được tiền xử lí sẵn nằm trong file dữ liệu thời tiết, lí do lựa chọn là vì file có sẵn dữ liệu trạm thời tiết trong đó, rất tiện cho việc join bảng sau này

In [45]:
path3 = 'weather/air_store_info_with_nearest_active_station.csv'
air_store_info = pd.read_csv(path3)
air_store_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 829 entries, 0 to 828
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   air_store_id          829 non-null    object 
 1   air_genre_name        829 non-null    object 
 2   air_area_name         829 non-null    object 
 3   latitude              829 non-null    float64
 4   longitude             829 non-null    float64
 5   latitude_str          829 non-null    object 
 6   longitude_str         829 non-null    object 
 7   station_id            829 non-null    object 
 8   station_latitude      829 non-null    float64
 9   station_longitude     829 non-null    float64
 10  station_vincenty      829 non-null    float64
 11  station_great_circle  829 non-null    float64
dtypes: float64(6), object(6)
memory usage: 77.8+ KB


- Thông qua việc đọc file `air_store_info_with_nearest_active_station.csv`, ta thấy được file có 11 cột như sau `air_store_id`, `air_genre_name`, `air_area_name`, `latitude`, `longitude`, `latitude_str`, `longitude_str`, `station_id`, `station_latitude`, `station_longitude`, `station_vincenty`, `station_great_circle`
- Ý nghĩa của các cột dữ liệu:
    - Với các cột `air_genre_name`, `air_area_name`:
        - Đây là các cột dữ liệu thể hiện loại hình, địa điểm của các nhà hàng
        - Trạng thái: các cột có 829 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: đều mang kiểu dữ liệu object
    - Với các cột `latitude`, `longitude`, `station_latitude`, `station_longitude`, `station_vincenty` và `station_great_circle`:
        - Đây là các cột dữ liệu `latitude`, `longitude` thể hiện kinh độ, vĩ độ của các cửa hàng, `station_latitude`, `station_longitude`, `station_vincenty` và `station_great_circle` thể hiện kinh độ, vĩ độ của các trạm thời tiết, `station_vincenty`, `station_great_circle` thể hiện khoảng cách từ nhà hàng đến trạm thời tiết gần đó theo ước lượng và tính chính xác
        - Trạng thái: các cột đều có 829 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: đều mang kiểu dữ liệu float64
    - Với các cột `langitude_str`, `longitude_str`, `station_id`:
        - Đây là các cột thể hiện kinh độ, vĩ độ dưới dạng string và id của trạm thời tiết gần đó
        - Trạng thái: các cột đều có 829 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: đều mang kiểu dữ liệu object

In [46]:
air_store_info.head(10)

air_store_id  air_genre_name                 air_area_name  \
0  air_0f0cdeee6c9bf3d7  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
1  air_7cc17a324ae5c7dc  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
2  air_fee8dcf4d619598e  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
3  air_a17f0778617c76e2  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
4  air_83db5aff8f50478e  Italian/French  Tōkyō-to Minato-ku Shibakōen   
5  air_99c3eae84130c1cb  Italian/French  Tōkyō-to Minato-ku Shibakōen   
6  air_f183a514cb8ff4fa  Italian/French  Tōkyō-to Minato-ku Shibakōen   
7  air_6b9fa44a9cf504a1  Italian/French  Tōkyō-to Minato-ku Shibakōen   
8  air_0919d54f0c9a24b8  Italian/French  Tōkyō-to Minato-ku Shibakōen   
9  air_2c6c79d597e48096  Italian/French  Tōkyō-to Minato-ku Shibakōen   

    latitude   longitude  latitude_str  longitude_str  \
0  34.695124  135.197853  "34.6951242"  "135.1978525"   
1  34.695124  135.197853  "34.6951242"  "135.1978525"   
2  34.695124  135.197853  "34.6951242"  "135.1978525"   
3  34.695124  135.197853  "34.6951242"  "135.1978525"   
4  35.658068  139.751599  "35.6580681"  "139.7515992"   
5  35.658068  139.751599  "35.6580681"  "139.7515992"   
6  35.658068  139.751599  "35.6580681"  "139.7515992"   
7  35.658068  139.751599  "35.6580681"  "139.7515992"   
8  35.658068  139.751599  "35.6580681"  "139.7515992"   
9  35.658068  139.751599  "35.6580681"  "139.7515992"   

                   station_id  station_latitude  station_longitude  \
0     hyogo__kobe-kana__koube         34.696667         135.211667   
1     hyogo__kobe-kana__koube         34.696667         135.211667   
2     hyogo__kobe-kana__koube         34.696667         135.211667   
3     hyogo__kobe-kana__koube         34.696667         135.211667   
4  tokyo__tokyo-kana__tonokyo         35.691667         139.750000   
5  tokyo__tokyo-kana__tonokyo         35.691667         139.750000   
6  tokyo__tokyo-kana__tonokyo         35.691667         139.750000   
7  tokyo__tokyo-kana__tonokyo         35.691667         139.750000   
8  tokyo__tokyo-kana__tonokyo         35.691667         139.750000   
9  tokyo__tokyo-kana__tonokyo         35.691667         139.750000   

   station_vincenty  station_great_circle  
0          1.277232              1.274882  
1          1.277232              1.274882  
2          1.277232              1.274882  
3          1.277232              1.274882  
4          3.730672              3.739835  
5          3.730672              3.739835  
6          3.730672              3.739835  
7          3.730672              3.739835  
8          3.730672              3.739835  
9          3.730672              3.739835

#### 2.1.4 `File submission.csv`

In [47]:
submission = pd.read_csv('data/sample_submission.csv')
submission.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32019 entries, 0 to 32018
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        32019 non-null  object
 1   visitors  32019 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 500.4+ KB


- File `submission` là file kết quả chúng ta sẽ nộp trên Kaggle còn được xem là file test sau này
- Thông qua việc đọc`submission.csv`, file chứa 2 cột là `id`, `visitors`
- Ý nghĩa của các cột dữ liệu:
    -  Với cột dữ liệu `id`:
        - Đây là từ khóa bao gồm id của cửa hàng + ngày có khách, thông qua từ khóa sẽ dự đoán ra số khách tương ứng
        - Trạng thái: cột id có 32019 dòng dữ liệu, và không trống
        - Kiểu dữ liệu: cột `id` có kiểu dữ liệu object

- **Ý tưởng xử lí file `submission`**:
    - Lấy cột id của cửa hàng, ngày có khách: Thông qua str.slice() để trích xuất từ cột `id`, với 21 kí tự đầu là id của cửa hàng, từ kí tự thứ 21 trở đi là ngày có khách
    - Đánh dấu tập test: tạo thêm 1 cột `is_test` với giá trị là True
    - Đánh số theo index cho từng dòng

In [48]:

submission['air_store_id'] = submission['id'].str.slice(0,20)
submission['visit_date'] = submission['id'].str.slice(21)
submission['visit_date'] = pd.to_datetime(submission['visit_date']).dt.strftime('%Y-%m-%d')
submission['visitors'] = np.nan #để bỏ qua cột này
submission['is_test'] = True
submission['is_number'] = submission.index

submission.head()


id  visitors          air_store_id  \
0  air_00a91d42b08b08d9_2017-04-23       NaN  air_00a91d42b08b08d9   
1  air_00a91d42b08b08d9_2017-04-24       NaN  air_00a91d42b08b08d9   
2  air_00a91d42b08b08d9_2017-04-25       NaN  air_00a91d42b08b08d9   
3  air_00a91d42b08b08d9_2017-04-26       NaN  air_00a91d42b08b08d9   
4  air_00a91d42b08b08d9_2017-04-27       NaN  air_00a91d42b08b08d9   

   visit_date  is_test  is_number  
0  2017-04-23     True          0  
1  2017-04-24     True          1  
2  2017-04-25     True          2  
3  2017-04-26     True          3  
4  2017-04-27     True          4

##

#### 2.1.5 Merge dataset

Mục đích merge để có tập dữ liệu đầy đủ và hoàn thiện hơn cho mô hình
- **Ý tưởng xử lí sau khi merge:**
    - Thay những giá trị Nan trong cột `is_test` bằng False, giải thích về điều này là vì sau khi ghép 2 bảng lại thì những dữ liệu trong file test thì giá trị cột `is_test` là True trong khi những cột khác lại là Nan nên ta mới có cách điều chỉnh như trên
    - Chuyển kiểu dữ liệu của cột `visitors` nhằm giữ lại các Nan là các giá trị của cột `visitors` nằm trong tập Test

In [49]:
data = pd.concat((air_visit, submission.drop(['id'], axis = 'columns')))
data['is_test']  = data['is_test'].fillna(False).astype('bool') #vì cột khi ép có type là object, dùng astype để chuyển sang boolean
data = pd.merge(left = data, right = date_info, on = 'visit_date', how = 'left')
data = pd.merge(left = data, right = air_store_info, on = 'air_store_id', how = 'left')
data['visitors'] = data['visitors'].astype(float)
data.head()

air_store_id  visit_date  visitors was_nil  is_test  is_number  \
0  air_00a91d42b08b08d9  2016-07-01      35.0   False    False        NaN   
1  air_00a91d42b08b08d9  2016-07-02       9.0   False    False        NaN   
2  air_00a91d42b08b08d9  2016-07-03       0.0    True    False        NaN   
3  air_00a91d42b08b08d9  2016-07-04      20.0   False    False        NaN   
4  air_00a91d42b08b08d9  2016-07-05      25.0   False    False        NaN   

  day_of_week  is_holiday  prev_holiday  after_holiday  ...  \
0      Friday           0           0.0            0.0  ...   
1    Saturday           0           0.0            0.0  ...   
2      Sunday           0           0.0            0.0  ...   
3      Monday           0           0.0            0.0  ...   
4     Tuesday           0           0.0            0.0  ...   

                     air_area_name   latitude   longitude  latitude_str  \
0  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595  "35.6940027"   
1  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595  "35.6940027"   
2  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595  "35.6940027"   
3  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595  "35.6940027"   
4  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595  "35.6940027"   

   longitude_str                  station_id station_latitude  \
0  "139.7535951"  tokyo__tokyo-kana__tonokyo        35.691667   
1  "139.7535951"  tokyo__tokyo-kana__tonokyo        35.691667   
2  "139.7535951"  tokyo__tokyo-kana__tonokyo        35.691667   
3  "139.7535951"  tokyo__tokyo-kana__tonokyo        35.691667   
4  "139.7535951"  tokyo__tokyo-kana__tonokyo        35.691667   

   station_longitude  station_vincenty  station_great_circle  
0             139.75          0.416011              0.415906  
1             139.75          0.416011              0.415906  
2             139.75          0.416011              0.415906  
3             139.75          0.416011              0.415906  
4             139.75          0.416011              0.415906  

[5 rows x 21 columns]

## 2.2. Dữ liệu thời tiết

- **Input:** ta sẽ có dữ liệu 1663 trạm đo thời tiết
- **Output:** ta sẽ có file dữ liệu bao gồm tất cả 1663 trạm đo 

In [2]:
import glob
from pathlib import Path
weather_df = []
for path in glob.glob('weather/1-1-16_5-31-17_Weather/1-1-16_5-31-17_Weather/*.csv'):
    weather = pd.read_csv(path)
    weather['station_id'] = Path(path).stem
    weather_df.append(weather)


weather = pd.concat(weather_df, axis='rows')
weather.rename(columns={'calendar_date': 'visit_date'}, inplace=True)


In [5]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
Index: 859771 entries, 0 to 516
Data columns (total 16 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   visit_date          859771 non-null  object 
 1   avg_temperature     480234 non-null  float64
 2   high_temperature    480189 non-null  float64
 3   low_temperature     480189 non-null  float64
 4   precipitation       640279 non-null  float64
 5   hours_sunlight      433473 non-null  float64
 6   solar_radiation     24803 non-null   float64
 7   deepest_snowfall    123827 non-null  float64
 8   total_snowfall      120157 non-null  float64
 9   avg_wind_speed      473007 non-null  float64
 10  avg_vapor_pressure  80383 non-null   float64
 11  avg_local_pressure  81048 non-null   float64
 12  avg_humidity        80384 non-null   float64
 13  avg_sea_pressure    78054 non-null   float64
 14  cloud_cover         30412 non-null   float64
 15  station_id          859771 non-null  objec

- Thông qua việc đọc file tất cả các file .csv trong weather, ta thấy được có 16 cột bao gồm `visit_date`, `avg_temperature`, `high_temperature`, `low_temperature`, `precipitation`, `hours_sunlight`, `solar_radiation`, `deepest_snowfall`, `total_snowfall`, `avg_wind_speed`, `avg_vapor_pressure`, `avg_local_pressure`, `avg_humidity`, `avg_sea_pressure`, `cloud_cover` và `station_id`:
- Nhìn qua tổng quan thì file có rất nhiều cột bị null, chỉ có cột lượng mưa `precipitation` và nhiệt độ trung bình `avg_temperature` thì ổn hơn và có vẻ dùng được
- Giải thích vì lí do chỉ chọn lượng mưa và nhiệt độ trung bình là vì:
    - `high_temperature` và `low_temperature` chỉ cho ta thấy được nhiệt độ cao nhất hoặc nhiệt độ thấp nhất trong ngày nhưng thứ ta cần lại là cái tổng quan hơn nên `avg_temperature` là lựa chọn hoàn hảo hơn
    - Còn lại đa số dữ liệu thiếu quá nhiều nên không thể dùng được, chỉ có lượng mưa thì đầy đủ nhất nên chúng ta sẽ chọn lượng mưa
- Còn về việc vì sao lại có việc dữ liệu lại thiếu nhiều như vậy, đó là bởi vì:
    - 

### 2.2.2 Trích xuất dữ liệu thời tiết

- **Input:** dữ liệu từ bảng weather ta đã tạo trước đó
- **Output:** trích xuất ra những thuộc tính cần thiết để phù hợp với mô hình
- **Ý tưởng:** Vì sao lại có việc trích xuất ở đây?
    * Đó là bởi vì số khách hàng của chúng ta sẽ tính theo ngày nên việc thuộc tính `high_temperature` và `low_temperature` sẽ không giúp ích được gì nhiều, chỉ cần `avg_temperature` là đủ
    * Một số thuộc tính thì lại không cần thiết hoặc do missing quá nhiều nên chúng ta bỏ qua
** Tóm lại chúng ta sẽ chọn 2 thuộc tính: `avg_temperature` và `percitation`**

In [52]:
means = weather.groupby('visit_date')[['avg_temperature', 'precipitation']].mean().reset_index()
means.rename(columns={'avg_temperature': 'global_avg_temperature', 'precipitation': 'global_precipitation'}, inplace=True)
weather = pd.merge(left=weather, right=means, on='visit_date', how='left')
weather['avg_temperature'].fillna(weather['global_avg_temperature'], inplace=True)
weather['precipitation'].fillna(weather['global_precipitation'], inplace=True)

weather[['visit_date', 'avg_temperature', 'precipitation']].head()

visit_date  avg_temperature  precipitation
0  2016-01-01              6.0            0.0
1  2016-01-02              4.7            0.0
2  2016-01-03              7.0            0.0
3  2016-01-04              8.8            0.0
4  2016-01-05              8.9            0.0

In [53]:
data['visit_date'] = pd.to_datetime(data['visit_date'])
data.index = data['visit_date']
# data.sort_values(['air_store_id', 'visit_date'], inplace=True)

data.head()

air_store_id visit_date  visitors was_nil  is_test  \
visit_date                                                               
2016-07-01  air_00a91d42b08b08d9 2016-07-01      35.0   False    False   
2016-07-02  air_00a91d42b08b08d9 2016-07-02       9.0   False    False   
2016-07-03  air_00a91d42b08b08d9 2016-07-03       0.0    True    False   
2016-07-04  air_00a91d42b08b08d9 2016-07-04      20.0   False    False   
2016-07-05  air_00a91d42b08b08d9 2016-07-05      25.0   False    False   

            is_number day_of_week  is_holiday  prev_holiday  after_holiday  \
visit_date                                                                   
2016-07-01        NaN      Friday           0           0.0            0.0   
2016-07-02        NaN    Saturday           0           0.0            0.0   
2016-07-03        NaN      Sunday           0           0.0            0.0   
2016-07-04        NaN      Monday           0           0.0            0.0   
2016-07-05        NaN     Tuesday           0           0.0            0.0   

            ...                    air_area_name   latitude   longitude  \
visit_date  ...                                                           
2016-07-01  ...  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595   
2016-07-02  ...  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595   
2016-07-03  ...  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595   
2016-07-04  ...  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595   
2016-07-05  ...  Tōkyō-to Chiyoda-ku Kudanminami  35.694003  139.753595   

            latitude_str  longitude_str                  station_id  \
visit_date                                                            
2016-07-01  "35.6940027"  "139.7535951"  tokyo__tokyo-kana__tonokyo   
2016-07-02  "35.6940027"  "139.7535951"  tokyo__tokyo-kana__tonokyo   
2016-07-03  "35.6940027"  "139.7535951"  tokyo__tokyo-kana__tonokyo   
2016-07-04  "35.6940027"  "139.7535951"  tokyo__tokyo-kana__tonokyo   
2016-07-05  "35.6940027"  "139.7535951"  tokyo__tokyo-kana__tonokyo   

           station_latitude  station_longitude  station_vincenty  \
visit_date                                                         
2016-07-01        35.691667             139.75          0.416011   
2016-07-02        35.691667             139.75          0.416011   
2016-07-03        35.691667             139.75          0.416011   
2016-07-04        35.691667             139.75          0.416011   
2016-07-05        35.691667             139.75          0.416011   

            station_great_circle  
visit_date                        
2016-07-01              0.415906  
2016-07-02              0.415906  
2016-07-03              0.415906  
2016-07-04              0.415906  
2016-07-05              0.415906  

[5 rows x 21 columns]

## 2.3. Tiền xử lí dữ liệu

### 2.2.1 Xử lí giá trị ngoại lai

- **Phương pháp:** Tính toán tứ phân phân vị thứ nhất Q1 và tứ phân vị thử ba Q3, khoảng tứ phân vị IQR. Sau đó nhóm lọc những outlier < Q1 - 1.5IQR hoặc outlier > Q3 + 1.5*IQR. Sau đó, ta sẽ thay những giá trị outlier bằng giá trị lớn của dữ liệu không nằm trong outlier
- **Mục tiêu:** 

In [57]:
def findOutlier(data):
    return (data - data.mean()) > 2.4 * data.std()
def replace_outlier(data):
    outlier = findOutlier(data)
    max = data[~outlier].max()
    data[outlier] = max
    return data

store = data.groupby('air_store_id')
data['outlier'] = store.apply(lambda g: findOutlier(g['visitors'])).values
data['replace_visitors'] = store.apply(lambda g: replace_outlier(g['visitors'])).values
data['replace_visitor_log1'] = np.log1p(data['replace_visitors'])
data.head()

air_store_id visit_date  visitors was_nil  is_test  \
visit_date                                                               
2016-07-01  air_00a91d42b08b08d9 2016-07-01      35.0   False    False   
2016-07-02  air_00a91d42b08b08d9 2016-07-02       9.0   False    False   
2016-07-03  air_00a91d42b08b08d9 2016-07-03       0.0    True    False   
2016-07-04  air_00a91d42b08b08d9 2016-07-04      20.0   False    False   
2016-07-05  air_00a91d42b08b08d9 2016-07-05      25.0   False    False   

            is_number day_of_week  is_holiday  prev_holiday  after_holiday  \
visit_date                                                                   
2016-07-01        NaN      Friday           0           0.0            0.0   
2016-07-02        NaN    Saturday           0           0.0            0.0   
2016-07-03        NaN      Sunday           0           0.0            0.0   
2016-07-04        NaN      Monday           0           0.0            0.0   
2016-07-05        NaN     Tuesday           0           0.0            0.0   

            ...                  station_id station_latitude  \
visit_date  ...                                                
2016-07-01  ...  tokyo__tokyo-kana__tonokyo        35.691667   
2016-07-02  ...  tokyo__tokyo-kana__tonokyo        35.691667   
2016-07-03  ...  tokyo__tokyo-kana__tonokyo        35.691667   
2016-07-04  ...  tokyo__tokyo-kana__tonokyo        35.691667   
2016-07-05  ...  tokyo__tokyo-kana__tonokyo        35.691667   

            station_longitude  station_vincenty station_great_circle outlier  \
visit_date                                                                     
2016-07-01             139.75          0.416011             0.415906   False   
2016-07-02             139.75          0.416011             0.415906   False   
2016-07-03             139.75          0.416011             0.415906   False   
2016-07-04             139.75          0.416011             0.415906   False   
2016-07-05             139.75          0.416011             0.415906   False   

           replace_visitors  replace_visitor_log1  weekend  day_of_month  
visit_date                                                                
2016-07-01             35.0              3.583519        0             1  
2016-07-02              9.0              2.302585        1             2  
2016-07-03              0.0              0.000000        1             3  
2016-07-04             20.0              3.044522        0             4  
2016-07-05             25.0              3.258097        0             5  

[5 rows x 26 columns]

### 2.2.2 Trích xuất các đặc trưng khác

In [58]:
data['weekend'] = data['day_of_week'].isin(['Saturday', 'Sunday']).astype(int)
data['day_of_month'] = data['visit_date'].dt.day

### 2.2.3 Trọng số theo cấp số nhân cho số lượng khách hàng

In [59]:
from scipy import optimize
def cal_ewm(data, alpha, adjust = False):
    return data.shift().ewm(alpha = alpha, adjust = adjust).mean()
def find_best_alpha(data, adjust = False, eps = 10e-5):
    def f(alpha):
        ewm_estimate = cal_ewm(data, alpha = min(max(alpha, 0), 1), adjust= adjust)
        error = np.mean((data - ewm_estimate)**2)
        return error
    best = optimize.differential_evolution(func = f, bounds = [(0 + eps, 1 - eps)])
    return cal_ewm(data, alpha = best['x'][0], adjust = adjust)

group = data.groupby(['air_store_id', 'day_of_week']).apply(lambda g: find_best_alpha(g['replace_visitors']))
data['optimize_week_visitor'] = group.sort_index(level=['air_store_id', 'visit_date']).values

group = data.groupby(['air_store_id', 'day_of_week']).apply(lambda g: find_best_alpha(g['replace_visitor_log1']))
data['optimize_week_visitor_log1'] = group.sort_index(level = ['air_store_id', 'visit_date']).values

group = data.groupby(['air_store_id', 'weekend']).apply(lambda g: find_best_alpha(g['replace_visitors']))
data['optimize_weekend_visitor']= group.sort_index(level= ['air_store_id', 'visit_date']).values

group = data.groupby(['air_store_id', 'weekend']).apply(lambda g: find_best_alpha(g['replace_visitor_log1']))
data['optimize_weekend_visitor_log1p'] = group.sort_index(level = ['air_store_id', 'visit_date']).values


data.head()

air_store_id visit_date  visitors was_nil  is_test  \
visit_date                                                               
2016-07-01  air_00a91d42b08b08d9 2016-07-01      35.0   False    False   
2016-07-02  air_00a91d42b08b08d9 2016-07-02       9.0   False    False   
2016-07-03  air_00a91d42b08b08d9 2016-07-03       0.0    True    False   
2016-07-04  air_00a91d42b08b08d9 2016-07-04      20.0   False    False   
2016-07-05  air_00a91d42b08b08d9 2016-07-05      25.0   False    False   

            is_number day_of_week  is_holiday  prev_holiday  after_holiday  \
visit_date                                                                   
2016-07-01        NaN      Friday           0           0.0            0.0   
2016-07-02        NaN    Saturday           0           0.0            0.0   
2016-07-03        NaN      Sunday           0           0.0            0.0   
2016-07-04        NaN      Monday           0           0.0            0.0   
2016-07-05        NaN     Tuesday           0           0.0            0.0   

            ... station_great_circle outlier  replace_visitors  \
visit_date  ...                                                  
2016-07-01  ...             0.415906   False              35.0   
2016-07-02  ...             0.415906   False               9.0   
2016-07-03  ...             0.415906   False               0.0   
2016-07-04  ...             0.415906   False              20.0   
2016-07-05  ...             0.415906   False              25.0   

            replace_visitor_log1 weekend day_of_month optimize_week_visitor  \
visit_date                                                                    
2016-07-01              3.583519       0            1                   NaN   
2016-07-02              2.302585       1            2                   NaN   
2016-07-03              0.000000       1            3                   NaN   
2016-07-04              3.044522       0            4                   NaN   
2016-07-05              3.258097       0            5                   NaN   

            optimize_week_visitor_log1  optimize_weekend_visitor  \
visit_date                                                         
2016-07-01                         NaN                       NaN   
2016-07-02                         NaN                       NaN   
2016-07-03                         NaN                   9.00000   
2016-07-04                         NaN                  35.00000   
2016-07-05                         NaN                  33.04746   

            optimize_weekend_visitor_log1p  
visit_date                                  
2016-07-01                             NaN  
2016-07-02                             NaN  
2016-07-03                        2.302585  
2016-07-04                        3.583519  
2016-07-05                        3.454852  

[5 rows x 30 columns]

#### Tính các đặc trưng khác

In [60]:
def rollStatistic(data, col, group_by):
    data.index.name = None
    data.sort_values(group_by + ['visit_date'], inplace = True)
    data_group = data.groupby(group_by, sort = False)
    stats = {
        'mean' : [],
        'median' : [],
        'std' : [],
        'count' : [],
        'max' : [],
        'min' : []
    }
    alpha_exp = [0.1, 0.25, 0.3, 0.5, 0.75]
    stats.update({'ewm_{}_mean'.format(alpha) : [] for alpha in alpha_exp})
    for _, group in data_group:
        shift = group[col].shift()
        rolling = shift.rolling(window = len(group), min_periods = 1)
        stats['mean'].extend(rolling.mean())
        stats['median'].extend(rolling.median())
        stats['std'].extend(rolling.std())
        stats['count'].extend(rolling.count())
        stats['max'].extend(rolling.max())
        stats['min'].extend(rolling.min())
        for alpha in alpha_exp:
            ewm_alpha = shift.ewm(alpha = alpha, adjust = False)
            stats['ewm_{}_mean'.format(alpha)].extend(ewm_alpha.mean())
    merge_name = '_&_'.join(group_by)
    for stats_name, value in stats.items():
        data['{}_{}_{}'.format(col, stats_name, merge_name)] = value

rollStatistic(data, col = 'replace_visitors', group_by = ['air_store_id', 'day_of_week'])
rollStatistic(data, col = 'replace_visitors', group_by = ['air_store_id', 'weekend'])
rollStatistic(data, col = 'replace_visitors', group_by = ['air_store_id'])

rollStatistic(data, col = 'replace_visitor_log1', group_by = ['air_store_id', 'day_of_week'])
rollStatistic(data, col = 'replace_visitor_log1', group_by = ['air_store_id', 'weekend'])
rollStatistic(data, col = 'replace_visitor_log1', group_by = ['air_store_id'])
data.sort_values(['air_store_id', 'visit_date'])
    
        

air_store_id visit_date  visitors was_nil  is_test  \
2016-07-01  air_00a91d42b08b08d9 2016-07-01      35.0   False    False   
2016-07-02  air_00a91d42b08b08d9 2016-07-02       9.0   False    False   
2016-07-03  air_00a91d42b08b08d9 2016-07-03       0.0    True    False   
2016-07-04  air_00a91d42b08b08d9 2016-07-04      20.0   False    False   
2016-07-05  air_00a91d42b08b08d9 2016-07-05      25.0   False    False   
...                          ...        ...       ...     ...      ...   
2017-05-27  air_fff68b929994bfbd 2017-05-27       NaN     NaN     True   
2017-05-28  air_fff68b929994bfbd 2017-05-28       NaN     NaN     True   
2017-05-29  air_fff68b929994bfbd 2017-05-29       NaN     NaN     True   
2017-05-30  air_fff68b929994bfbd 2017-05-30       NaN     NaN     True   
2017-05-31  air_fff68b929994bfbd 2017-05-31       NaN     NaN     True   

            is_number day_of_week  is_holiday  prev_holiday  after_holiday  \
2016-07-01        NaN      Friday           0           0.0            0.0   
2016-07-02        NaN    Saturday           0           0.0            0.0   
2016-07-03        NaN      Sunday           0           0.0            0.0   
2016-07-04        NaN      Monday           0           0.0            0.0   
2016-07-05        NaN     Tuesday           0           0.0            0.0   
...               ...         ...         ...           ...            ...   
2017-05-27    32014.0    Saturday           0           0.0            0.0   
2017-05-28    32015.0      Sunday           0           0.0            0.0   
2017-05-29    32016.0      Monday           0           0.0            0.0   
2017-05-30    32017.0     Tuesday           0           0.0            0.0   
2017-05-31    32018.0   Wednesday           0           0.0            1.0   

            ... replace_visitor_log1_median_air_store_id  \
2016-07-01  ...                                      NaN   
2016-07-02  ...                                 3.583519   
2016-07-03  ...                                 2.943052   
2016-07-04  ...                                 2.302585   
2016-07-05  ...                                 2.673554   
...         ...                                      ...   
2017-05-27  ...                                 3.401197   
2017-05-28  ...                                 3.401197   
2017-05-29  ...                                 3.401197   
2017-05-30  ...                                 3.401197   
2017-05-31  ...                                 3.401197   

           replace_visitor_log1_std_air_store_id  \
2016-07-01                                   NaN   
2016-07-02                                   NaN   
2016-07-03                              0.905757   
2016-07-04                              1.815870   
2016-07-05                              1.578354   
...                                          ...   
2017-05-27                              1.221014   
2017-05-28                              1.221014   
2017-05-29                              1.221014   
2017-05-30                              1.221014   
2017-05-31                              1.221014   

            replace_visitor_log1_count_air_store_id  \
2016-07-01                                      0.0   
2016-07-02                                      1.0   
2016-07-03                                      2.0   
2016-07-04                                      3.0   
2016-07-05                                      4.0   
...                                             ...   
2017-05-27                                    296.0   
2017-05-28                                    296.0   
2017-05-29                                    296.0   
2017-05-30                                    296.0   
2017-05-31                                    296.0   

            replace_visitor_log1_max_air_store_id  \
2016-07-01                                    NaN   
2016-07-02                               3.583519   
2016-07-03                  

In [ ]:
print(data.columns.tolist())

['air_store_id', 'visit_date', 'visitors', 'was_nil', 'is_test', 'is_number', 'day_of_week', 'is_holiday', 'prev_holiday', 'after_holiday', 'air_genre_name', 'air_area_name', 'latitude', 'longitude', 'latitude_str', 'longitude_str', 'station_id', 'station_latitude', 'station_longitude', 'station_vincenty', 'station_great_circle', 'outlier', 'replace_visitors', 'replace_visitor_log1', 'weekend', 'day_of_month', 'optimize_week_visitor', 'optimize_week_visitor_log1', 'optimize_weekend_visitor', 'optimize_weekend_visitor_log1p', 'replace_visitors_mean_air_store_id_&_day_of_week', 'replace_visitors_median_air_store_id_&_day_of_week', 'replace_visitors_std_air_store_id_&_day_of_week', 'replace_visitors_count_air_store_id_&_day_of_week', 'replace_visitors_max_air_store_id_&_day_of_week', 'replace_visitors_min_air_store_id_&_day_of_week', 'replace_visitors_ewm_0.1_mean_air_store_id_&_day_of_week', 'replace_visitors_ewm_0.25_mean_air_store_id_&_day_of_week', 'replace_visitors_ewm_0.3_mean_air_s

Huấn luyện mô hình

Mã hóa dữ liệu

Lí do vì sao dùng one-hot encoding 

In [61]:
data = pd.get_dummies(data, columns=['day_of_week', 'air_genre_name'])
data.head()

air_store_id visit_date  visitors was_nil  is_test  \
2016-07-01  air_00a91d42b08b08d9 2016-07-01      35.0   False    False   
2016-07-02  air_00a91d42b08b08d9 2016-07-02       9.0   False    False   
2016-07-03  air_00a91d42b08b08d9 2016-07-03       0.0    True    False   
2016-07-04  air_00a91d42b08b08d9 2016-07-04      20.0   False    False   
2016-07-05  air_00a91d42b08b08d9 2016-07-05      25.0   False    False   

            is_number  is_holiday  prev_holiday  after_holiday  \
2016-07-01        NaN           0           0.0            0.0   
2016-07-02        NaN           0           0.0            0.0   
2016-07-03        NaN           0           0.0            0.0   
2016-07-04        NaN           0           0.0            0.0   
2016-07-05        NaN           0           0.0            0.0   

                              air_area_name  ...  air_genre_name_Dining bar  \
2016-07-01  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   
2016-07-02  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   
2016-07-03  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   
2016-07-04  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   
2016-07-05  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   

            air_genre_name_International cuisine  \
2016-07-01                                 False   
2016-07-02                                 False   
2016-07-03                                 False   
2016-07-04                                 False   
2016-07-05                                 False   

           air_genre_name_Italian/French air_genre_name_Izakaya  \
2016-07-01                          True                  False   
2016-07-02                          True                  False   
2016-07-03                          True                  False   
2016-07-04                          True                  False   
2016-07-05                          True                  False   

           air_genre_name_Japanese food  air_genre_name_Karaoke/Party  \
2016-07-01                        False                         False   
2016-07-02                        False                         False   
2016-07-03                        False                         False   
2016-07-04                        False                         False   
2016-07-05                        False                         False   

            air_genre_name_Okonomiyaki/Monja/Teppanyaki  air_genre_name_Other  \
2016-07-01                                        False                 False   
2016-07-02                                        False                 False   
2016-07-03                                        False                 False   
2016-07-04                                        False                 False   
2016-07-05                                        False                 False   

            air_genre_name_Western food  air_genre_name_Yakiniku/Korean food  
2016-07-01                        False                                False  
2016-07-02                        False                                False  
2016-07-03                        False                                False  
2016-07-04                        False                                False  
2016-07-05                        False                                False  

[5 rows x 115 columns]

In [ ]:
data.loc[data['is_test'] == True].head()

air_store_id visit_date  visitors was_nil  is_test  \
2017-04-23  air_00a91d42b08b08d9 2017-04-23       NaN     NaN     True   
2017-04-24  air_00a91d42b08b08d9 2017-04-24       NaN     NaN     True   
2017-04-25  air_00a91d42b08b08d9 2017-04-25       NaN     NaN     True   
2017-04-26  air_00a91d42b08b08d9 2017-04-26       NaN     NaN     True   
2017-04-27  air_00a91d42b08b08d9 2017-04-27       NaN     NaN     True   

            is_number  is_holiday  prev_holiday  after_holiday  \
2017-04-23        0.0           0           0.0            0.0   
2017-04-24        1.0           0           0.0            0.0   
2017-04-25        2.0           0           0.0            0.0   
2017-04-26        3.0           0           0.0            0.0   
2017-04-27        4.0           0           0.0            0.0   

                              air_area_name  ...  air_genre_name_Dining bar  \
2017-04-23  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   
2017-04-24  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   
2017-04-25  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   
2017-04-26  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   
2017-04-27  Tōkyō-to Chiyoda-ku Kudanminami  ...                      False   

            air_genre_name_International cuisine  \
2017-04-23                                 False   
2017-04-24                                 False   
2017-04-25                                 False   
2017-04-26                                 False   
2017-04-27                                 False   

           air_genre_name_Italian/French air_genre_name_Izakaya  \
2017-04-23                          True                  False   
2017-04-24                          True                  False   
2017-04-25                          True                  False   
2017-04-26                          True                  False   
2017-04-27                          True                  False   

           air_genre_name_Japanese food  air_genre_name_Karaoke/Party  \
2017-04-23                        False                         False   
2017-04-24                        False                         False   
2017-04-25                        False                         False   
2017-04-26                        False                         False   
2017-04-27                        False                         False   

            air_genre_name_Okonomiyaki/Monja/Teppanyaki  air_genre_name_Other  \
2017-04-23                                        False                 False   
2017-04-24                                        False                 False   
2017-04-25                                        False                 False   
2017-04-26                                        False                 False   
2017-04-27                                        False                 False   

            air_genre_name_Western food  air_genre_name_Yakiniku/Korean food  
2017-04-23                        False                                False  
2017-04-24                        False                                False  
2017-04-25                        False                                False  
2017-04-26                        False                                False  
2017-04-27                        False                                False  

[5 rows x 115 columns]

In [62]:
data['visitor_log1p'] = np.log1p(data['visitors'])
train = data[(data['is_test'] == False) & (data['outlier'] == False) & (data['was_nil'] == False)]
test = data[data['is_test'] == True].sort_values('is_number')
#liet ke cac cot de xoa
col_drop = ['air_store_id', 'visitors', 'is_test', 'outlier', 'visit_date', 'was_nil', 'is_number',
            'replace_visitors', 'air_area_name', 'station_id', 'station_latitude', 'station_longitude', 'station_vincenty',
            'station_great_circle', 'replace_visitor_log1', 'latitude_str', 'longitude_str']

train = train.drop(col_drop, axis = 'columns')
train = train.dropna()
X_train = train.drop('visitor_log1p', axis = 'columns')
Y_train = train['visitor_log1p']

test = test.drop(col_drop, axis = 'columns')
X_test = test.drop('visitor_log1p', axis = 'columns')
X_test.head()

is_holiday  prev_holiday  after_holiday   latitude   longitude  \
2017-04-23           0           0.0            0.0  35.694003  139.753595   
2017-04-24           0           0.0            0.0  35.694003  139.753595   
2017-04-25           0           0.0            0.0  35.694003  139.753595   
2017-04-26           0           0.0            0.0  35.694003  139.753595   
2017-04-27           0           0.0            0.0  35.694003  139.753595   

            weekend  day_of_month  optimize_week_visitor  \
2017-04-23        1            23              24.246640   
2017-04-24        0            24              19.781789   
2017-04-25        0            25               2.913289   
2017-04-26        0            26              16.033537   
2017-04-27        0            27              12.523343   

            optimize_week_visitor_log1  optimize_weekend_visitor  ...  \
2017-04-23                    3.068320                 16.230913  ...   
2017-04-24                    2.477768                 23.341027  ...   
2017-04-25                    1.324953                 13.520350  ...   
2017-04-26                    2.312020                 11.247147  ...   
2017-04-27                    2.314172                  8.513239  ...   

            air_genre_name_Dining bar  air_genre_name_International cuisine  \
2017-04-23                      False                                 False   
2017-04-24                      False                                 False   
2017-04-25                      False                                 False   
2017-04-26                      False                                 False   
2017-04-27                      False                                 False   

            air_genre_name_Italian/French  air_genre_name_Izakaya  \
2017-04-23                           True                   False   
2017-04-24                           True                   False   
2017-04-25                           True                   False   
2017-04-26                           True                   False   
2017-04-27                           True                   False   

            air_genre_name_Japanese food  air_genre_name_Karaoke/Party  \
2017-04-23                         False                         False   
2017-04-24                         False                         False   
2017-04-25                         False                         False   
2017-04-26                         False                         False   
2017-04-27                         False                         False   

            air_genre_name_Okonomiyaki/Monja/Teppanyaki  air_genre_name_Other  \
2017-04-23                                        False                 False   
2017-04-24                                        False                 False   
2017-04-25                                        False                 False   
2017-04-26                                        False                 False   
2017-04-27                                        False                 False   

            air_genre_name_Western food  air_genre_name_Yakiniku/Korean food  
2017-04-23                        False                                False  
2017-04-24                        False                                False  
2017-04-25                        False                                False  
2017-04-26                        False                                False  
2017-04-27                        False                                False  

[5 rows x 98 columns]

LightGBM + KFold

In [63]:


np.random.seed(42)

# --- model ---
model = lgb.LGBMRegressor(
    objective='regression',
    max_depth=5,
    num_leaves=5 ** 2 - 1,
    learning_rate=0.007,
    n_estimators=30000,
    min_child_samples=80,
    subsample=0.8,
    colsample_bytree=1,
    reg_alpha=0,
    reg_lambda=0,
    random_state=np.random.randint(10e6),
    verbosity=-1
)

n_splits = 6
cv = KFold(n_splits=n_splits, shuffle=True, random_state=42)

val_scores = [0] * n_splits

sub = submission['id'].to_frame()
sub['visitors'] = 0

feature_importances = pd.DataFrame(index=X_train.columns)

for i, (fit_idx, val_idx) in enumerate(cv.split(X_train, Y_train)):

    X_fit = X_train.iloc[fit_idx]
    y_fit = Y_train.iloc[fit_idx]
    X_val = X_train.iloc[val_idx]
    y_val = Y_train.iloc[val_idx]

    # --- tạo progress bar ---
    pbar = tqdm(total=model.n_estimators, desc=f'Fold {i+1}', ncols=100)

    def tqdm_callback(env):
        """Callback cho tqdm, gọi mỗi iteration"""
        pbar.update(1)

    # train model với callback
    model.fit(
        X_fit,
        y_fit,
        eval_set=[(X_val, y_val)],
        eval_names=['val'],
        eval_metric='l2',
        feature_name=X_fit.columns.tolist(),
        callbacks=[tqdm_callback]
    )

    pbar.close()

    val_scores[i] = np.sqrt(model.best_score_['val']['l2'])
    sub['visitors'] += model.predict(X_test, num_iteration=model.best_iteration_)
    feature_importances[i] = model.feature_importances_

# trung bình kết quả
sub['visitors'] /= n_splits
sub['visitors'] = np.expm1(sub['visitors'])

val_mean = np.mean(val_scores)
val_std = np.std(val_scores)

print('Local RMSLE: {:.5f} (±{:.5f})'.format(val_mean, val_std))


Fold 6: 100%|████████████████████████████████████████████████| 30000/30000 [02:48<00:00, 178.06it/s]


Local RMSLE: 0.54544 (±0.00197)


In [ ]:
sub.to_csv('submissionslgbm_{:.5f}_{:.5f}.csv'.format(val_mean, val_std), index=False)

In [ ]:
csv = pd.read_csv('submissionslgbm_0.54679_0.00300.csv')
csv['visitors'] = csv['visitors'].round().astype(int)
csv.head()

id  visitors
0  air_00a91d42b08b08d9_2017-04-23        15
1  air_00a91d42b08b08d9_2017-04-24        21
2  air_00a91d42b08b08d9_2017-04-25        26
3  air_00a91d42b08b08d9_2017-04-26        30
4  air_00a91d42b08b08d9_2017-04-27        27